Core/MIPS/IR/IRInst.cpp

```diff
@@ -0,0 +1,589 @@
+			MIPSInterpret(op);
+			break;
+		}
+
+		case IROp::CallReplacement:
+		{
+			int funcIndex = constPool[inst->src1];
+			const ReplacementTableEntry *f = GetReplacementFunc(funcIndex);
+			int cycles = f->replaceFunc();
+			mips->downcount -= cycles;
+			return mips->r[MIPS_REG_RA];
+		}
+
+		case IROp::Break:
+			Crash();
+			break;
+
+		case IROp::SetCtrlVFPU:
+			mips->vfpuCtrl[inst->dest] = constPool[inst->src1];
+			break;
+
+		default:
+			Crash();
+		}
+#ifdef _DEBUG
+		if (mips->r[0] != 0)
+			Crash();
+#endif
+		inst++;
+	}
+
+	// If we got here, the block was badly constructed.
+	Crash();
+	return 0;
+}
+
+void IRWriter::Write(IROp op, u8 dst, u8 src1, u8 src2) {
+	IRInst inst;
+	inst.op = op;
+	inst.dest = dst;
+	inst.src1 = src1;
+	inst.src2 = src2;
+	insts_.push_back(inst);
+}
+
+void IRWriter::WriteSetConstant(u8 dst, u32 value) {
+	Write(IROp::SetConst, dst, AddConstant(value));
+}
+
+int IRWriter::AddConstant(u32 value) {
+	for (size_t i = 0; i < constPool_.size(); i++) {
+		if (constPool_[i] == value)
+			return (int)i;
+	}
+	constPool_.push_back(value);
+	if (constPool_.size() > 255) {
+		// Cannot have more than 256 constants in a block!
+		Crash();
+	}
+	return (int)constPool_.size() - 1;
+}
+
+int IRWriter::AddConstantFloat(float value) {
+	u32 val;
+	memcpy(&val, &value, 4);
+	return AddConstant(val);
+}
+
+void IRWriter::Simplify() {
+	SimplifyInPlace(&insts_[0], (int)insts_.size(), constPool_.data());
+}
+
+const char *GetGPRName(int r) {
+	if (r < 32) {
+		return currentDebugMIPS->GetRegName(0, r);
+	}
+	switch (r) {
+	case IRTEMP_0: return "irtemp0";
+	case IRTEMP_1: return "irtemp1";
+	case IRTEMP_LHS: return "irtemp_lhs";
+	case IRTEMP_RHS: return "irtemp_rhs";
+	default: return "(unk)";
+	}
+}
+
+void DisassembleParam(char *buf, int bufSize, u8 param, char type, const u32 *constPool) {
+	switch (type) {
+	case 'G':
+		snprintf(buf, bufSize, "%s", GetGPRName(param));
+		break;
+	case 'F':
+		snprintf(buf, bufSize, "r%d", param);
+		break;
+	case 'C':
+		snprintf(buf, bufSize, "%08x", constPool[param]);
+		break;
+	case 'I':
+		snprintf(buf, bufSize, "%02x", param);
+		break;
+	case '_':
+	case '\0':
+		buf[0] = 0;
+		break;
+	default:
+		snprintf(buf, bufSize, "?");
+		break;
+	}
+}
+
+const IRMeta *GetIRMeta(IROp op) {
+	return metaIndex[(int)op];
+}
+
+void DisassembleIR(char *buf, size_t bufsize, IRInst inst, const u32 *constPool) {
+	const IRMeta *meta = GetIRMeta(inst.op);
+	if (!meta) {
+		snprintf(buf, bufsize, "Unknown %d", (int)inst.op);
+		return;
+	}
+	char bufDst[16];
+	char bufSrc1[16];
+	char bufSrc2[16];
+	DisassembleParam(bufDst, sizeof(bufDst) - 2, inst.dest, meta->types[0], constPool);
+	DisassembleParam(bufSrc1, sizeof(bufSrc1) - 2, inst.src1, meta->types[1], constPool);
+	DisassembleParam(bufSrc2, sizeof(bufSrc2), inst.src2, meta->types[2], constPool);
+	if (meta->types[1] && meta->types[0] != '_') {
+		strcat(bufDst, ", ");
+	}
+	if (meta->types[2] && meta->types[1] != '_') {
```

Comment:
it's defined as a 4 char array ... so I hope the rest get 0'd. Not sure though..